File: Sem6/JPP/Lista4/CPP/zad2.cpp
#include <iostream>
#include <thread>
#include <vector>
#include <queue>
#include <mutex>
#include <condition_variable>
#include <random>
#include <chrono>
#include <atomic>
#include <deque>
#include <sstream>

using namespace std;

struct Message {
    int from;
    int to;
    int msg_id;
};

void safe_print(const string& msg) {
    static mutex cout_mtx;
    lock_guard<mutex> lock(cout_mtx);
    cout << msg << endl;
}

class User {
public:
    int id;
    queue<Message> incoming;
    mutex mtx;
    condition_variable cv;
    int sent_count = 0;
    int received_count = 0;

    User(int i) : id(i) {}
    
    User(const User&) = delete;
    User& operator=(const User&) = delete;
};

class Server {
private:
    deque<User>& users;
    mutex mtx;
    condition_variable cv_to_server;  
    condition_variable cv_from_server; 
    
    bool busy = false;
    bool msg_available = false;
    Message current_msg;
    atomic<bool> running{true};

public:
    Server(deque<User>& u) : users(u) {}

    void send_to_server(Message msg) {
        unique_lock<mutex> lock(mtx);
        
        cv_to_server.wait(lock, [this] { return !busy || !running; });
        if (!running) return;

        current_msg = msg;
        msg_available = true;
        busy = true;

        stringstream ss;
        ss << "[SERVER] Odebrano od U" << msg.from << " dla U" << msg.to << " (ID: " << msg.msg_id << ")";
        safe_print(ss.str());
        
        cv_from_server.notify_one();
        
        cv_to_server.wait(lock, [this] { return !busy || !running; });
    }

    void run() {
        while (running) {
            unique_lock<mutex> lock(mtx);
            cv_from_server.wait(lock, [this] { return msg_available || !running; });

            if (!running) break;

            Message msg = current_msg;
            msg_available = false;

            {
                lock_guard<mutex> user_lock(users[msg.to].mtx);
                users[msg.to].incoming.push(msg);
                users[msg.to].received_count++;
                users[msg.to].cv.notify_one();
            }

            stringstream ss;
            ss << "[SERVER] Wiadomość ID:" << msg.msg_id << " dostarczona do U" << msg.to;
            safe_print(ss.str());

            busy = false; 
            cv_to_server.notify_all(); 
        }
    }

    void stop() {
        running = false;
        cv_to_server.notify_all();
        cv_from_server.notify_all();
    }
};

void sender_func(User& me, Server& server, int num_msg, int num_users) {
    mt19937 gen(me.id + time(0));
    uniform_int_distribution<> dist(0, num_users - 1);

    for (int i = 0; i < num_msg; ++i) {
        int recipient = dist(gen);
        Message msg{me.id, recipient, i};
        
        stringstream ss;
        ss << "[U" << me.id << "] Próbuje wysłać wiadomość ID:" << i << " do U" << recipient;
        safe_print(ss.str());

        server.send_to_server(msg);
        me.sent_count++;
        
        this_thread::sleep_for(chrono::milliseconds(20)); 
    }
}

void receiver_func(User& me) {
    while (true) {
        unique_lock<mutex> lock(me.mtx);
        if (me.cv.wait_for(lock, chrono::seconds(2), [&me] { return !me.incoming.empty(); })) {
            while (!me.incoming.empty()) {
                Message msg = me.incoming.front();
                me.incoming.pop();
                
                stringstream ss;
                ss << "         [U" << me.id << "] OTRZYMAŁEM wiadomość ID:" << msg.msg_id << " od U" << msg.from;
                safe_print(ss.str());
            }
        } else {
            break;
        }
    }
}

int main(int argc, char* argv[]) {
    int num_users = 5;
    int messages_per_user = 5;

    if (argc == 3) {
        num_users = stoi(argv[1]);
        messages_per_user = stoi(argv[2]);
    } else {
        cout << "Użycie: " << argv[0] << " <użytkownicy> <wiadomości>" << endl;
        cout << "Uruchamiam z domyślnymi parametrami: 5 użytkowników, 5 wiadomości.\n" << endl;
    }

    deque<User> users;
    for (int i = 0; i < num_users; ++i) users.emplace_back(i);

    Server server(users);
    thread s_thread(&Server::run, &server);

    vector<thread> senders;
    vector<thread> receivers;

    for (int i = 0; i < num_users; ++i) {
        senders.emplace_back(sender_func, ref(users[i]), ref(server), messages_per_user, num_users);
        receivers.emplace_back(receiver_func, ref(users[i]));
    }

    for (auto& t : senders) t.join();
    
    server.stop();
    s_thread.join();
    for (auto& t : receivers) t.join();

    cout << "STATYSTYKI KOŃCOWE\n";
    
    int total_sent = 0;
    int total_received = 0;

    for (const auto& u : users) {
        cout << "Użytkownik " << u.id << ": Wysłano " << u.sent_count 
             << " | Odebrano " << u.received_count << endl;
        total_sent += u.sent_count;
        total_received += u.received_count;
    }

    cout << "\nŁącznie wysłanych:  " << total_sent << endl;
    cout << "Łącznie odebranych: " << total_received << endl;

    if (total_sent == total_received) {
        cout << "SUKCES: Wszystkie komunikaty dotarły." << endl;
    } else {
        cout << "BŁĄD: Wykryto stratę komunikatów!" << endl;
    }

    return 0;
}

File: Sem6/JPP/Lista4/GO/zad2.go
package main

import (
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"sync"
	"time"
)

type Message struct {
	from   int
	to     int
	msgID  int
}

type User struct {
	id            int
	inbox         chan Message
	sentCount     int
	receivedCount int
	done          chan bool
}

type Server struct {
	users []*User
}

func (s *Server) Transfer(msg Message) {
	fmt.Printf("[SERVER] Przesyłam: U%d -> U%d (ID: %d)\n", msg.from, msg.to, msg.msgID)
	
	s.users[msg.to].inbox <- msg
	
	fmt.Printf("[SERVER] Dostarczono do U%d\n", msg.to)
}

func main() {
	numUsers := 0
	msgPerUser := 0

	if len(os.Args) == 3 {
		u, err1 := strconv.Atoi(os.Args[1])
		m, err2 := strconv.Atoi(os.Args[2])
		if err1 != nil || err2 != nil {
			fmt.Println("Błąd: Argumenty muszą być liczbami całkowitymi.")
			os.Exit(1)
		}
		numUsers = u
		msgPerUser = m
	} else {
		fmt.Println("Użycie: program <liczba_użytkowników> <wiadomości_na_użytkownika>")
		os.Exit(1)
	}

	fmt.Printf("System: %d użytkowników, %d wiadomości na każdego\n\n", numUsers, msgPerUser)

	users := make([]*User, numUsers)
	for i := 0; i < numUsers; i++ {
		users[i] = &User{
			id:    i,
			inbox: make(chan Message), 
			done:  make(chan bool),
		}
	}

	server := &Server{users: users}
	
	serverMtx := sync.Mutex{}

	var wg sync.WaitGroup

	for i := 0; i < numUsers; i++ {
		wg.Add(1)
		go func(u *User) {
			defer wg.Done()
			timeout := time.After(2 * time.Second)
			for {
				select {
				case msg := <-u.inbox:
					u.receivedCount++
					fmt.Printf("         [U%d] Odebrałem ID:%d od U%d\n", u.id, msg.msgID, msg.from)
					timeout = time.After(2 * time.Second)
				case <-timeout:
					return
				}
			}
		}(users[i])
	}

	for i := 0; i < numUsers; i++ {
		wg.Add(1)
		go func(u *User) {
			defer wg.Done()
			r := rand.New(rand.NewSource(time.Now().UnixNano() + int64(u.id)))
			
			for m := 0; m < msgPerUser; m++ {
				target := r.Intn(numUsers)
				msg := Message{from: u.id, to: target, msgID: m}
				
				fmt.Printf("[U%d] Wysyłam ID:%d do U%d\n", u.id, m, target)
				
				serverMtx.Lock()
				server.Transfer(msg)
				serverMtx.Unlock()
				
				u.sentCount++
				time.Sleep(time.Millisecond * 20)
			}
		}(users[i])
	}

	wg.Wait()

	fmt.Println("STATYSTYKI KOŃCOWE")
	totalSent, totalReceived := 0, 0
	for _, u := range users {
		fmt.Printf("Użytkownik %d: Wysłano %d | Odebrano %d\n", u.id, u.sentCount, u.receivedCount)
		totalSent += u.sentCount
		totalReceived += u.receivedCount
	}
	fmt.Printf("\nŁącznie wysłano: %d\nŁącznie odebrano: %d\n", totalSent, totalReceived)
}